pkg/apis/rollouts/validation/validation.go

```diff
@@ -0,0 +1,268 @@
+	ScaleDownDelayLongerThanDeadlineMessage = "ScaleDownDelaySeconds cannot be longer than ProgressDeadlineSeconds"
+	// RolloutMinReadyLongerThanDeadlineMessage indicates the MinReadySeconds is longer than ProgressDeadlineSeconds
+	RolloutMinReadyLongerThanDeadlineMessage = "MinReadySeconds cannot be longer than ProgressDeadlineSeconds"
+	// InvalidStrategyMessage indiciates that multiple strategies can not be listed
+	InvalidStrategyMessage = "Multiple Strategies can not be listed"
+	// DuplicatedServicesMessage the message to indicate that the rollout uses the same service for the active and preview services
+	DuplicatedServicesMessage = "This rollout uses the same service for the active and preview services, but two different services are required."
+	// ScaleDownLimitLargerThanRevisionLimit the message to indicate that the rollout's revision history limit can not be smaller than the rollout's scale down limit
+	ScaleDownLimitLargerThanRevisionLimit = "This rollout's revision history limit can not be smaller than the rollout's scale down limit"
+	// AvailableReason the reason to indicate that the rollout is serving traffic from the active service
+	AvailableReason = "AvailableReason"
+	// NotAvailableMessage the message to indicate that the Rollout does not have min availability
+	NotAvailableMessage = "Rollout does not have minimum availability"
+	// AvailableMessage the message to indicate that the Rollout does have min availability
+	AvailableMessage = "Rollout has minimum availability"
+)
+
+// called in strategy.go -> create.go
+func ValidateRollout(rollout *v1alpha1.Rollout) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateRolloutSpec(rollout, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateRolloutSpec checks for a valid spec otherwise returns a list of erros.
+func ValidateRolloutSpec(rollout *v1alpha1.Rollout, fldPath *field.Path) field.ErrorList {
+	spec := rollout.Spec
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(*spec.Replicas), fldPath.Child("replicas"))...)
+
+	if spec.Selector == nil {
+		message := fmt.Sprintf(MissingFieldMessage, ".Spec.Selector")
+		allErrs = append(allErrs, field.Required(fldPath.Child("selector"), message))
+	} else {
+		allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(spec.Selector, fldPath.Child("selector"))...)
+		if len(spec.Selector.MatchLabels)+len(spec.Selector.MatchExpressions) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "empty selector is invalid for deployment"))
+		}
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(spec.Selector)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "invalid label selector"))
+	} else {
+		data, _ := json.Marshal(&spec.Template)
+		var template core.PodTemplateSpec
+		_ = json.Unmarshal(data, &template)
```

Comment:
We should get the err and add it to the error list if it fails instead of silently failing.